main/lsp/lsp.h

```diff
@@ -201,59 +182,25 @@ class LSPLoop {
-                                       UnorderedMap<std::string, SorbetWorkspaceFileUpdate> &updates) const;
-    void preprocessSorbetWorkspaceEdit(const WatchmanQueryResponse &queryResponse,
-                                       UnorderedMap<std::string, SorbetWorkspaceFileUpdate> &updates) const;
-    TypecheckRun handleSorbetWorkspaceEdit(std::unique_ptr<core::GlobalState> gs,
-                                           const DidChangeTextDocumentParams &changeParams) const;
-    TypecheckRun handleSorbetWorkspaceEdit(std::unique_ptr<core::GlobalState> gs,
-                                           const DidOpenTextDocumentParams &openParams) const;
-    TypecheckRun handleSorbetWorkspaceEdit(std::unique_ptr<core::GlobalState> gs,
-                                           const DidCloseTextDocumentParams &closeParams) const;
-    TypecheckRun handleSorbetWorkspaceEdit(std::unique_ptr<core::GlobalState> gs,
-                                           const WatchmanQueryResponse &queryResponse) const;
-    TypecheckRun handleSorbetWorkspaceEdits(std::unique_ptr<core::GlobalState> gs,
-                                            std::vector<std::unique_ptr<SorbetWorkspaceEdit>> &edits) const;
-    TypecheckRun commitSorbetWorkspaceEdits(std::unique_ptr<core::GlobalState> gs,
-                                            UnorderedMap<std::string, SorbetWorkspaceFileUpdate> &updates) const;
-    static std::string_view getFileContents(UnorderedMap<std::string, LSPLoop::SorbetWorkspaceFileUpdate> &updates,
-                                            const core::GlobalState &initialGS, std::string_view path);
+    TypecheckRun handleSorbetWorkspaceEdits(std::unique_ptr<core::GlobalState> gs, LSPFileUpdates &updates) const;
 
     /** Returns `true` if 5 minutes have elapsed since LSP last sent counters to statsd. */
     bool shouldSendCountersToStatsd(std::chrono::time_point<std::chrono::steady_clock> currentTime) const;
     /** Sends counters to statsd. */
     void sendCountersToStatsd(std::chrono::time_point<std::chrono::steady_clock> currentTime);
 
 public:
-    LSPLoop(std::unique_ptr<core::GlobalState> gs, LSPConfiguration config, const std::shared_ptr<spd::logger> &logger,
-            WorkerPool &workers, int inputFd, std::ostream &output);
-    std::unique_ptr<core::GlobalState> runLSP();
-    LSPResult processRequest(std::unique_ptr<core::GlobalState> gs, const LSPMessage &msg);
+    LSPLoop(std::unique_ptr<core::GlobalState> initialGS, LSPConfiguration config,
+            const std::shared_ptr<spd::logger> &logger, WorkerPool &workers, int inputFd, std::ostream &output);
+    /**
+     * Runs the language server on a dedicated thread. Returns the final global state if it exits cleanly, or nullopt
+     * on error.
+     */
+    std::optional<std::unique_ptr<core::GlobalState>> runLSP();
+    LSPResult processRequest(std::unique_ptr<core::GlobalState> gs, std::unique_ptr<LSPMessage> msg);
```

Comment:
One unfortunate casuality: I couldn't easily keep the `LSPMessage` a const ref. All messages must be preprocessed through the preprocessor now, which has an interface that moves messages into a queue.

I don't think we give up too much by making it non-const. If this turns out to be a mistake, we can see about changing how LSPLoop and LSPPreprocessor interact.